### PR TITLE
The patch to docker cli package needed to make `docker compose` work

### PR DESCRIPTION
**Pull request by mike-sul**
I've just created a pull request out of the Andy's commit, which is basically copy-paste of the patch to `docker cli` needed for building of `docker` itself https://github.com/foundriesio/meta-lmp/blob/main/meta-lmp-base/recipes-containers/docker/files/cli-config-support-default-system-config.patch.

Signed-off-by: Andy Doan <andy@foundries.io>

### CODE REVIEW

**Comment by mike-sul** [comment]
I verified it long time ago within my factory and QEMU device, will double check today.

**Comment by mike-sul** [comment]
Works fine for the simple Compose App. Tested against the latest aklite that uses `docker compose`.

File: meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc [hunk 1]
```diff
@@ -3,7 +3,7 @@ CORE_IMAGE_BASE_INSTALL += " \
     docker \
     docker-credential-helper-fio \
     python3-docker \
-    python3-docker-compose \
+    docker-compose \
```
Comment (doanac) [inline meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: i think ricardo will want this done in a separate patch. that just slipped in during my debugging.

Comment (mike-sul) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: ok. So, we need both `python3-docker-compose` and `docker-compose` in this PR?

Comment (doanac) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: I think we'll switch to just "docker-compose" eventually (during this release). Not sure if we want to make that switch *now*?

Comment (ricardosalveti) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: Yeah, please drop this one from this PR. We can do the switch in another pr.

Comment (mike-sul) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: If I drop change to this file then just `python3-docker-compose` is installed. Then it's not clear how installation of the golang `docker-compose` that is patched in this PR is triggered at all.
So, I am asking whether we need both `python3-docker-compose` and `docker-compose` included?

Comment (doanac) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc:6]: this PR is for building a correct docker-compose in factories that explicitly enable it. people can then build this in their factory if they really want it. then in another PR or commit (probably a PR) we do the switch from python to golang.

Comment (mike-sul) [reply meta-lmp-base/recipes-samples/images/lmp-feature-docker.inc]: updated

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb [hunk 2]
```diff
@@ -20,6 +26,9 @@ GO_EXTRA_LDFLAGS = "-w -X ${GO_IMPORT}/internal.Version=${PV}"
 go_do_compile() {
 	export TMPDIR="${GOTMPDIR}"
 	mkdir -p ${B}/cli-plugins/bin
+	${GO} mod download -modcacherw
+	chmod -R +w ${B}/pkg/mod
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:30]: Do we need chmod with modcacherw?

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb [hunk 3]
```diff
@@ -20,6 +26,9 @@ GO_EXTRA_LDFLAGS = "-w -X ${GO_IMPORT}/internal.Version=${PV}"
...
+	cp ${S}/src/github.com/docker/cli/cli/config/config.go ${B}/pkg/mod/github.com/docker/cli@v20.10.3-0.20220309205733-2b52f62e9627+incompatible/cli/config/config.go
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:31]: @doanac can we create a link for the entire folder here?

File: meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch [hunk 4]
```diff
@@ -0,0 +1,41 @@
+From 15e560830117514adc825c80326e7db66192e717 Mon Sep 17 00:00:00 2001
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch:1]: Can you have this patch under the docker-compose/docker-compose folder instead of docker-compose/files?

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb (hunk 2 above)
Comment (ricardosalveti) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:30]:
```
-       chmod -R +w ${B}/pkg/mod
-       cp ${S}/src/github.com/docker/cli/cli/config/config.go ${B}/pkg/mod/github.com/docker/cli@v20.10.3-0.20220309205733-2b52f62e9627+incompatible/cli/config/config.go
+       cp -f ${S}/src/github.com/docker/cli/cli/config/config.go ${B}/pkg/mod/github.com/docker/cli@v20.10.3-0.20220309205733-2b52f62e9627+incompatible/cli/config/config.go
```

Seems to work ^^

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb (hunk 3 above)
Comment (ricardosalveti) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:31]: Don't worry about this one, this gets way too complicated quite quickly :-)

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb (hunk 2 above)
Comment (mike-sul) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:30]: I guess `cp` will fail without `modcacherw`

File: meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch (hunk 4 above)
Comment (mike-sul) [reply meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch]: ok

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb (hunk 2 above)
Comment (ricardosalveti) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb]: I tested with cp -f and it worked fine locally.

Comment (mike-sul) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb]: Do you mean without `-modcacherw`?

**Comment by mike-sul** [comment]
@ricardosalveti make changes according to your suggestions. https://ci.foundries.io/projects/msul-dev01/lmp/builds/1301/

File: meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+@@ -94,6 +95,15 @@ func Load(configDir string) (*configfile.ConfigFile, error) {
+ 	filename := filepath.Join(configDir, ConfigFileName)
+ 	configFile := configfile.New(filename)
+
++	// LmP: Load values from system config by default
++	if _, err := os.Stat(defaultSystemConfig); err == nil {
++		file, err := os.Open(defaultSystemConfig)
++		if err == nil {
++			configFile.LoadFromReader(file)
++			file.Close()
```
[the remark this answers is not in the document]
Comment (doanac) [reply meta-lmp-base/recipes-containers/docker-compose/files/cli-config-support-default-system-config.patch]: i kind of dislike defers for one-liners like that. and - its a short lived process, we could leak everything and its not going to matter.

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb (hunk 2 above)
Comment (doanac) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb]: i was thinking it was needed for cleanup logic (rm was failing), but i might have just copied it from some other recipe without thinking it through.

**Review by doanac — COMMENTED** [review]
looks good to me but i'm a bad "baker"

File: meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb [hunk 6]
```diff
@@ -20,6 +26,8 @@ GO_EXTRA_LDFLAGS = "-w -X ${GO_IMPORT}/internal.Version=${PV}"
 go_do_compile() {
 	export TMPDIR="${GOTMPDIR}"
 	mkdir -p ${B}/cli-plugins/bin
+	${GO} mod download
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb:29]: We just need to add back '-modcacherw' here.

Comment (ricardosalveti) [reply meta-lmp-base/recipes-containers/docker-compose/docker-compose_2.4.1.bb]: The clean step will fail later on without it (and failed locally when I used this patch).

**Review by ricardosalveti — APPROVED** [review]
LGTM
